File: src/main/java/pl/allegro/experiments/chi/chiserver/domain/experiments/PermissionsRepository.java
package pl.allegro.experiments.chi.chiserver.domain.experiments;

import com.google.common.base.Preconditions;
import pl.allegro.experiments.chi.chiserver.domain.User;
import pl.allegro.experiments.chi.chiserver.domain.UserProvider;

public class PermissionsRepository {
    private final UserProvider userProvider;
    private final ExperimentsRepository experimentsRepository;

    public PermissionsRepository(
            UserProvider userProvider,
            ExperimentsRepository experimentsRepository) {
        Preconditions.checkNotNull(userProvider);
        Preconditions.checkNotNull(experimentsRepository);
        this.userProvider = userProvider;
        this.experimentsRepository = experimentsRepository;
    }

    public Experiment withPermissions(Experiment experiment) {
        User currentUser = userProvider.getCurrentUser();
        return experiment.withEditableFlag(currentUser.isOwner(experiment))
                .withOrigin(experimentsRepository.getOrigin(experiment.getId()));
    }
}

File: src/main/kotlin/pl/allegro/experiments/chi/chiserver/domain/experiments/PermissionsRepository.kt
package pl.allegro.experiments.chi.chiserver.domain.experiments

import pl.allegro.experiments.chi.chiserver.domain.UserProvider

class PermissionsRepository(private val userProvider: UserProvider,
                            private val experimentsRepository : ExperimentsRepository) {
    fun withPermissions(experiment: Experiment): Experiment {
        val currentUser = userProvider.getCurrentUser()
        return experiment.withEditableFlag(currentUser.isOwner(experiment))
                         .withOrigin(experimentsRepository.getOrigin(experiment.id))
    }
}